components/wifi/esp8266-driver/ESP8266Interface.cpp

```diff
@@ -184,13 +184,19 @@ void ESP8266Interface::_connect_async()
         return;
     }
     _connect_retval = _esp.connect(ap_ssid, ap_pass);
+    int timeleft_ms = ESP8266_INTERFACE_CONNECT_TIMEOUT_MS - _conn_timer.read_ms();
     if (_connect_retval == NSAPI_ERROR_OK || _connect_retval == NSAPI_ERROR_AUTH_FAILURE
-            || _connect_retval == NSAPI_ERROR_NO_SSID) {
+            || _connect_retval == NSAPI_ERROR_NO_SSID
+            || ((_if_blocking == true) && (timeleft_ms <= 0))) {
         _connect_event_id = 0;
+        _conn_timer.stop();
+        if (timeleft_ms <= 0) {
+            _connect_retval = NSAPI_ERROR_CONNECTION_TIMEOUT;
+        }
         _if_connected.notify_all();
     } else {
         // Postpone to give other stuff time to run
-        _connect_event_id = _global_event_queue->call_in(ESP8266_CONNECT_TIMEOUT, callback(this, &ESP8266Interface::_connect_async));
+        _connect_event_id = _global_event_queue->call_in(timeleft_ms, callback(this, &ESP8266Interface::_connect_async));
```

Comment:
This leads to a situation where connection procedure will be run twice at most in 45 second time period. Plusit it might take 15 more seconds to finish the last attempt -> returning to application might take up to ~60 seconds. Which is fine but should be somehow reflected on the documentation.

Reply:
Thank you for your patience with explaining this to me, @VeijoPesonen . I finally got what you meant. Indeed I did not consider the fact the the _connect_async contains a blocking call to _esp.connect, which might take a while.
I adjusted the documentation of the function to:
```
     *  If interface is configured blocking it will timeout after up to
     *  ESP8266_INTERFACE_CONNECT_TIMEOUT_MS + ESP8266_CONNECT_TIMEOUT ms.
```
and decreased the grace period, as you suggested to 5s. So right now we have:
2*15 + 5 + 15
So depending on the responsiveness of ESP, if it cannot connect and does not want to report any errors, connect will return after 35-50 seconds.

Reply:
No problem :) The code should also reflect the length of the grace period you proposed - now it just takes all the remaining time there is left. Please replace _timeleft_ms_ with the grace period. And if there is time left after trying twice we might as well continue trying until the whole window is used. I assume the modem might report an error before all 15 seconds gets used awaiting for an answer.

Reply:
I guess that's another valid point, I created
```
#define ESP8266_INTERFACE_CONNECT_INTERVAL_MS (5000)
```
as I though "grace period" might be too vague. I used this to establish the `_connect_async` callback period.
Just a thought for future - would it be acceptable to synchronise the ESP8266::connect and ESP8266Interface::_connect_async with some kind of semaphore instead of polling?